supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -0,0 +1,462 @@
+        method: str | None = None,
+        **target_options: Any,
+    ) -> None:
+        """Submit the circuit samples to the desired target device and store the resulting
+        probabilities.
+
+        The set of circuits is partitioned as necessary to not exceed the maximum circuits that can
+        be submitted to the given target device. The function then waits for the jobs to complete
+        before saving the resulting probability distributions.
+
+        Args:
+            target: The name of the target device.
+            shots: The number of shots to use. Defaults to 10,000
+            method: Optional method to use on the Superstaq device. Defaults to None corresponding
+                to normal running.
+            target_options: Optional configuration kwargs passed when submitting the job.
+        """
+        for sample in tqdm(self.samples):
+            if sample.job is not None:
+                continue
+            try:
+                job = self._service.create_job(
+                    sample.circuit,
+                    target=target,
+                    method=method,
+                    repetitions=shots,
+                    **target_options,
+                )
+                sample.job = job.job_id()
+            except SuperstaqServerException as error:
+                warnings.warn(
+                    "The following error ocurred when submitting the jobs to the server and not\n"
+                    "all jobs have been submitted. If this is a timeout or limit based error\n"
+                    "consider running `submit_ss_jobs(args)` again to continue submitting\n"
+                    "the outstanding jobs.\n"
+                    f"{error.message}"
+                )
+
+    def sample_statuses(self) -> list[str | None]:
+        """Returns:
+        Get the statuses of the jobs associated with each sample. If no job is associated
+        with a sample then :code:`None` is listed instead.
+        """
+        statuses: list[str | None] = []
+        for sample in self.samples:
+            if sample.job is None:
+                statuses.append(None)
+            else:
+                statuses.append(self._service.get_job(sample.job).status())
+        return statuses
+
+    def retrieve_ss_jobs(self) -> dict[str, str]:
+        """Retrieve the jobs from the server.
+
+        Returns:
+            A dictionary of the statuses of any jobs that have not been successfully completed.
+        """
+        # If no jobs then return empty dictionary
+        if not [sample for sample in self.samples if sample.job is not None]:
+            return {}
+
+        statuses = {}
+        waiting_samples = [
+            sample for sample in self.samples if not hasattr(sample, "probabilities")
+        ]
+        for sample in tqdm(waiting_samples, "Retrieving jobs"):
+            if sample.job is None:
+                continue
+            job = self._service.get_job(sample.job)
```

Comment:
jobs do some work to prevent unnecessary api calls/etc - it would be good to hold on to the job objects themselves instead of calling `get_job()` every time we need to check them